sklearn/metrics/cluster/unsupervised.py

```diff
@@ -102,6 +107,36 @@ def silhouette_score(X, labels, metric='euclidean', sample_size=None,
     return np.mean(silhouette_samples(X, labels, metric=metric, **kwds))
 
 
+def _silhouette_reduce(D_chunk, start, labels, label_freqs):
+    """Accumulate silhouette statistics for vertical chunk of X
+
+    Parameters
+    ----------
+    D_chunk : shape (n_chunk_samples, n_samples)
+        precomputed distances for a chunk
+    start : int
+        first index in chunk
+    labels : array, shape (n_samples,)
+        corresponding cluster labels, encoded as {0, ..., n_clusters-1}
+    label_freqs : array
+        distribution of cluster labels in ``labels``
+    """
+    # accumulate distances from each sample to each cluster
+    clust_dists = np.zeros((len(D_chunk), len(label_freqs)),
+                           dtype=D_chunk.dtype)
+    np.add.at(clust_dists.T, labels, D_chunk.T)
```

Comment:
There are some reports about `np.add.at` being slow (https://github.com/numpy/numpy/issues/5922,  https://github.com/numpy/numpy/issues/11156) and it's currently almost not used in the code base, do you have an opposite experience with it?

I guess there is not much alternatives since 
```py
clust_dists.T[labels] += D_chunk.T
```
and `np.bincount` (as suggested in the first issue) wouldn't work here as far as I understand?

Reply:
I'm unaware of those reports. I'll have a quick look.

Reply:
`clust_dists.T[labels] += D_chunk.T` will certainly not work if labels has duplicates, which it does. No, bincount won't work either, given that the output is 2d.

We could do this in a for loop. Is it worth trying that, or should we assume that it's not the bottleneck??

Reply:
Yes, I'm getting better mileage with a for loop (assuming I've correctly implemented it)

```py

n_samples = 10000
n_samples_chunk = 100
n_clusters = 10

D_chunk = np.random.rand(n_samples_chunk, n_samples)
labels = np.random.randint(n_clusters, size=n_samples)
label_freqs = np.bincount(labels)


clust_dists = np.zeros((len(D_chunk), len(label_freqs)),
                       dtype=D_chunk.dtype)
%timeit np.add.at(clust_dists.T, labels, D_chunk.T)

clust_dists = np.zeros((len(D_chunk), len(label_freqs)),
                       dtype=D_chunk.dtype)
%timeit for i in range(len(D_chunk)): clust_dists[i] += np.bincount(labels, weights=D_chunk[i], minlength=len(label_freqs))
```